tempodb/encoding/vparquet2/block_traceql.go

```diff
@@ -1779,41 +1775,39 @@ func (c *traceCollector) KeepGroup(res *parquetquery.IteratorResult) bool {
 			finalSpanset.StartTimeUnixNanos = e.Value.Uint64()
 		case columnPathDurationNanos:
 			finalSpanset.DurationNanos = e.Value.Uint64()
-			c.traceAttrs[traceql.NewIntrinsic(traceql.IntrinsicTraceDuration)] = traceql.NewStaticDuration(time.Duration(finalSpanset.DurationNanos))
+			c.traceAttrs[traceql.IntrinsicTraceDurationAttribute] = traceql.NewStaticDuration(time.Duration(finalSpanset.DurationNanos))
 		case columnPathRootSpanName:
 			finalSpanset.RootSpanName = e.Value.String()
-			c.traceAttrs[traceql.NewIntrinsic(traceql.IntrinsicTraceRootSpan)] = traceql.NewStaticString(finalSpanset.RootSpanName)
+			c.traceAttrs[traceql.IntrinsicTraceRootSpanAttribute] = traceql.NewStaticString(finalSpanset.RootSpanName)
 		case columnPathRootServiceName:
 			finalSpanset.RootServiceName = e.Value.String()
-			c.traceAttrs[traceql.NewIntrinsic(traceql.IntrinsicTraceRootService)] = traceql.NewStaticString(finalSpanset.RootServiceName)
+			c.traceAttrs[traceql.IntrinsicTraceRootServiceAttribute] = traceql.NewStaticString(finalSpanset.RootServiceName)
 		}
 	}
 
 	// Pre-allocate the final number of spans
 	numSpans := 0
 	for _, e := range res.OtherEntries {
-		if spanset, ok := e.Value.(*traceql.Spanset); ok {
-			numSpans += len(spanset.Spans)
+		if _, ok := e.Value.(*span); ok {
+			numSpans++
 		}
 	}
 	if cap(finalSpanset.Spans) < numSpans {
 		finalSpanset.Spans = make([]traceql.Span, 0, numSpans)
 	}
-
 	for _, e := range res.OtherEntries {
-		if spanset, ok := e.Value.(*traceql.Spanset); ok {
-			finalSpanset.Spans = append(finalSpanset.Spans, spanset.Spans...)
-
-			// loop over all spans and add the trace-level attributes
-			for k, v := range c.traceAttrs {
-				for _, sp := range spanset.Spans {
-					s := sp.(*span)
-					if _, alreadyExists := s.attributes[k]; !alreadyExists {
-						s.attributes[k] = v
-					}
-				}
+		if span, ok := e.Value.(*span); ok {
+			finalSpanset.Spans = append(finalSpanset.Spans, span)
+		}
+	}
+
+	// loop over all spans and add the trace-level attributes
+	for k, v := range c.traceAttrs {
+		for _, s := range finalSpanset.Spans {
+			s := s.(*span)
+			if _, alreadyExists := s.attributes[k]; !alreadyExists {
+				s.attributes[k] = v
```

Comment:
we do a lot of copying values onto the span struct into the one attributes map. i'm wondering if the span struct could contain multiple attribute maps for span/resource/trace/etc. levels. then we wouldn't need to copy the data. we could pass the same resource map to all attributes in the same batch.

the interface would have to change from this
```
func (s *span) Attributes() map[traceql.Attribute]traceql.Static {
	return s.attributes
}
```
to something like:
```
func (s *span) Attributes(traceqlAttribute) traceql.Static {
	 // check each map in precedence order
}
```

just an idle thought i had looking through the PR. not something for this PR, but do you think that would work?

Reply:
Yes that sounds like a neat improvement.  Could do the same for future things like Scopes.